addrman: improve performance by using more suitable containers

vasild opened the pull request:
`CAddrMan` uses `std::map` internally even though it does not require
that the map's elements are sorted. `std::map`'s access time is
`O(log(map size))`. `std::unordered_map` is more suitable as it has a
`O(1)` access time.

This patch lowers the execution times of `CAddrMan`'s methods as follows
(as per `src/bench/addrman.cpp`):

```
AddrMan::Add(): -3.5%
AddrMan::GetAddr(): -76%
AddrMan::Good(): -0.38%
AddrMan::Select(): -45%
```

vasild committed:
addrman: use unordered_map instead of map

`CAddrMan` uses `std::map` internally even though it does not require
that the map's elements are sorted. `std::map`'s access time is
`O(log(map size))`. `std::unordered_map` is more suitable as it has a
`O(1)` access time.

This patch lowers the execution times of `CAddrMan`'s methods as follows
(as per `src/bench/addrman.cpp`):

```
AddrMan::Add(): -3.5%
AddrMan::GetAddr(): -76%
AddrMan::Good(): -0.38%
AddrMan::Select(): -45%
```

vasild commented:
Full output from the benchmark:

```
map$ ./src/bench/bench_bitcoin -filter="AddrMan.*"
# Benchmark, evals, iterations, total, min, max, median
AddrManAdd, 5, 5, 4.45807, 0.176704, 0.179945, 0.178443
AddrManGetAddr, 5, 500, 4.09847, 0.00163663, 0.00164301, 0.0016392
AddrManGood, 5, 2, 4.66502, 0.460193, 0.474046, 0.467368
AddrManSelect, 5, 1000000, 2.07299, 4.12892e-07, 4.15564e-07, 4.147e-07

unordered_map$ ./src/bench/bench_bitcoin -filter="AddrMan.*"
# Benchmark, evals, iterations, total, min, max, median
AddrManAdd, 5, 5, 4.30032, 0.169753, 0.1738, 0.172047
AddrManGetAddr, 5, 500, 0.967398, 0.000384406, 0.00038996, 0.000387176
AddrManGood, 5, 2, 4.64742, 0.460599, 0.471216, 0.464143
AddrManSelect, 5, 1000000, 1.14085, 2.23899e-07, 2.36962e-07, 2.26767e-07
```

Empact commented on src/addrman.h:
```diff
@@ -181,14 +182,17 @@ friend class CAddrManTest;
 
 private:
+    typedef std::unordered_map<int, CAddrInfo> mapInfoType;
+    typedef std::unordered_map<CNetAddr, int, CNetAddrHash> mapAddrType;
```
nit: `using` and capitalize class

vasild replied:
Done: https://github.com/bitcoin/bitcoin/pull/18722/commits/0c3becc83f2cd53ffb5495430d27380f6eda248a#diff-0681caa5b164f2de0cf944039f84b19aR184-R185

Empact commented on src/addrman.h:
```diff
@@ -21,6 +21,7 @@
 #include <set>
 #include <stdint.h>
 #include <streams.h>
+#include <unordered_map>
```
nit: looks like you can move the `<map>` include to `addrman.cpp` now.

vasild replied:
Right! And even better - `std::map` is only used inside `CAddrMan::Check_()` in a way that fits `unordered_map` better - we only insert elements, query elements by key and check container size.

So I changed `CAddrMan::Check_()` to use `unordered_(set|map)` and removed the `#include <map>`.

MarcoFalke commented:
Would it be possible to submit the benchmark upfront, so that https://codespeed.bitcoinperf.com/ can start running it?

vasild commented:
Done: https://github.com/bitcoin/bitcoin/pull/18754

vasild commented:
Removed the first commit (which introduced the benchmark), now that it is already merged into `master` and rebased.

sipa commented on src/netaddress.h:
```diff
@@ -107,6 +107,19 @@ class CNetAddr
+        friend struct CNetAddrHash;
+};
+
+struct CNetAddrHash
```
You'll want to use a salted hash here (use CSipHasher from crypto/siphash.h or a specialization; see SaltedOutpointHasher in coins.h for inspiration). Otherwise an attacker may be able to degrade our performance by giving us a bunch of entries that all end up in the same bucket).

vasild replied:
Done, thanks for the suggestion!

sipa commented on src/addrman.h:
```diff
@@ -483,15 +486,28 @@ friend class CAddrManTest;
             }
         }
+#else // pre C++14, drop this once we switch to C++14 or higher
+        std::vector<MapInfo::const_iterator> toDelete;
```
Nit: new code should use the coding style described in doc/developer-.nodes.md (e.g. `to_delete` instead).

vasild replied:
Done

sipa commented:
utACK 7cc317285f388f0cda974a321761992248257fb8

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
Benchmarking results:
- master (8c97780db8c9dd33efed134385573ba97e9cd165)
```
$ ./src/bench/bench_bitcoin -filter="AddrMan.*"
# Benchmark, evals, iterations, total, min, max, median
AddrManAdd, 5, 5, 1.15889, 0.0458257, 0.0469068, 0.0464069
AddrManGetAddr, 5, 500, 1.94467, 0.000772949, 0.000789915, 0.000775272
AddrManGood, 5, 2, 5.69597, 0.563135, 0.57652, 0.570197
AddrManSelect, 5, 1000000, 1.17994, 2.32975e-07, 2.37942e-07, 2.36478e-07
```

- master + this PR
```
$ ./src/bench/bench_bitcoin -filter="AddrMan.*"
# Benchmark, evals, iterations, total, min, max, median
AddrManAdd, 5, 5, 0.972042, 0.0383488, 0.0396571, 0.0386554
AddrManGetAddr, 5, 500, 0.649199, 0.000256967, 0.000262957, 0.00025942
AddrManGood, 5, 2, 5.35751, 0.531294, 0.540723, 0.534681
AddrManSelect, 5, 1000000, 0.657766, 1.29313e-07, 1.36428e-07, 1.30873e-07
```

Nice :)

hebasto commented on src/netaddress.h:
```diff
@@ -107,6 +109,22 @@ class CNetAddr
+    {
+        CSipHasher hasher(salt_k0, salt_k1);
+        hasher.Write(a.ip, sizeof(a.ip));
+        return (size_t)hasher.Finalize();
```
Is this safe on 32-bit system (e.g., ARM7)?

vasild replied:
Yes, I think it is.

`Finalize()` returns `uint64_t` and even if `size_t` is 32 bits, then it will just take "half" of the hash. [`SaltedOutpointHasher`](https://github.com/bitcoin/bitcoin/blob/1c86ed4/src/coins.h#L108) already does such a conversion from `uint64_t` to `size_t` and it contains some comment that `size_t` must be returned.

sipa replied:
Any reason to not include `m_net` as well?

vasild replied:
No. That code is from before `m_net` existed! I added it now, thanks!

hebasto reviewed (commented):
nit: If `#include`s in `addrman.h` are already touched, mind moving
```
#include <fs.h>
#include <hash.h>
#include <streams.h>
```
into the first group?

vasild commented on src/addrman.cpp:
```diff
@@ -73,12 +76,12 @@ double CAddrInfo::GetChance(int64_t nNow) const
 CAddrInfo* CAddrMan::Find(const CNetAddr& addr, int* pnId)
 {
-    std::map<CNetAddr, int>::iterator it = mapAddr.find(addr);
+    auto it = mapAddr.find(addr);
```
Right, I will add the `const` if I end up modifying this PR for some bigger reason. But I think it is not worth invalidating the ACK.

vasild replied:
Done.

vasild commented on src/addrman.cpp:
```diff
@@ -73,12 +76,12 @@ double CAddrInfo::GetChance(int64_t nNow) const
     if (pnId)
         *pnId = (*it).second;
-    std::map<int, CAddrInfo>::iterator it2 = mapInfo.find((*it).second);
+    auto it2 = mapInfo.find((*it).second);
```
Right, I will add the `const` if I end up modifying this PR for some bigger reason. But I think it is not worth invalidating the ACK.

vasild replied:
Done.

vasild commented:
> nit: If #includes in addrman.h are already touched, mind moving...

Good catch, they should be moved! Given that there is already one ACK, I will move the headers if I alter this PR or as a followup PR if this gets merged as is now.

hebasto reviewed (approved):
ACK 7cc317285f388f0cda974a321761992248257fb8

@vasild feel free to ignore my nits :)

hebasto commented on src/netaddress.h:
```diff
@@ -107,6 +109,22 @@ class CNetAddr
+public:
+    size_t operator()(const CNetAddr& a) const noexcept
+    {
+        CSipHasher hasher(salt_k0, salt_k1);
+        hasher.Write(a.ip, sizeof(a.ip));
+        return (size_t)hasher.Finalize();
```
nit: All calls could be chained:
```suggestion
        return CSipHasher(salt_k0, salt_k1).Write(a.ip, sizeof(a.ip)).Finalize();
```

vasild replied:
I actually find the 3-line variant easier to read (that could be subjective). The 3-line variant is more manageable wrt setting a breakpoint or reading a backtrace (e.g. if you only see it crashed on that multi-chained-line it wouldn't be clear whether it crashed in the constructor, in the `Write()` method or in the `Finalize()` method).

I would prefer to leave it as is.

hebasto commented on src/netaddress.h:
```diff
@@ -107,6 +109,22 @@ class CNetAddr
+    }
+
+private:
+    const uint64_t salt_k0 = GetRand(std::numeric_limits<uint64_t>::max());
+    const uint64_t salt_k1 = GetRand(std::numeric_limits<uint64_t>::max());
```
nit: Class member naming convention?

vasild replied:
Right! A convention is a convention. Renamed.

vasild commented:
Prefixed `salt_k0` and `salt_k1` with `m_` to abide to the naming convention. [diff](https://github.com/bitcoin/bitcoin/compare/7cc317285f388f0cda974a321761992248257fb8..a72fee672a2da10a620136d3b3ddfb0336b95969).

hebasto commented:
@vasild 
> Prefixed `salt_k0` and `salt_k1` with `m_` to abide to the naming convention. [diff](https://github.com/bitcoin/bitcoin/compare/7cc317285f388f0cda974a321761992248257fb8..a72fee672a2da10a620136d3b3ddfb0336b95969).

If some changes are made already, maybe make these ones ((https://github.com/bitcoin/bitcoin/pull/18722#discussion_r440238143), https://github.com/bitcoin/bitcoin/pull/18722#discussion_r440237953, https://github.com/bitcoin/bitcoin/pull/18722#issuecomment-644188260) too?

:)

vasild commented:
I forgot about those, sorry. Done now.

Since the ACKs - renamed variables to follow convention, added `const`, moved `#include`s. [diff](https://github.com/bitcoin/bitcoin/compare/7cc317285f388f0cda974a321761992248257fb8..d6e782174)

hebasto reviewed (approved):
re-ACK d6e782174ecb7298e9d67303e52cc470ec9e2be7, only suggested changes since the [previous](https://github.com/bitcoin/bitcoin/pull/18722#pullrequestreview-430797229) review.

adamjonas commented:
Updated benchmarking results (now with [nanobench](https://github.com/martinus/nanobench), which meant I had to look up what these categories represented). In sum, it looks like a performance upgrade, especially for `AddrManGetAddr` though the error rate did jump up for me and I had to run the PR version a bunch of times to get something stable while master was stable the first time (though reading through [18011](https://github.com/bitcoin/bitcoin/pull/18011#issuecomment-643607657) this may be OS specific - this was run on OSX).

- master (609ce2d0da495ee09db03ba0c2d9b38a3b3f541e)
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       47,378,973.00 |               21.11 |    2.1% |      0.53 | `AddrManAdd`
|        1,585,044.00 |              630.90 |    1.8% |      0.02 | `AddrManGetAddr`
|      427,775,318.00 |                2.34 |    1.0% |      2.13 | `AddrManGood`
|              468.57 |        2,134,157.56 |    1.5% |      0.00 | `AddrManSelect`
```

- master + PR
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|       42,701,933.00 |               23.42 |    2.6% |      0.47 | `AddrManAdd`
|          572,564.00 |            1,746.53 |    3.8% |      0.01 | `AddrManGetAddr`
|      429,306,009.00 |                2.33 |    2.6% |      2.17 | `AddrManGood`
|              406.60 |        2,459,419.58 |    5.0% |      0.00 | `AddrManSelect`
```

vasild commented:
Rebased to resolve conflicts.

hebasto commented on src/netaddress.h:
```diff
@@ -12,6 +12,8 @@
 #include <attributes.h>
 #include <compat.h>
 #include <prevector.h>
+#include <crypto/siphash.h>
```
nit: The latest rebasing destroyed alphabetic order of headers :)

vasild replied:
Fixed, someday we may consider to have an automated clang-format enforcement in CI...

hebasto reviewed (approved):
re-ACK 1e1cc9d8594b0ffba1173d8e431e351228691536, only rebased.

hebasto reviewed (approved):
re-ACK 7dddf330fe9f17e20df7bc7431ad3f7b740a3a9c, only headers got correct order.

jonatack commented on src/netaddress.h:
```diff
@@ -190,6 +192,7 @@ class CNetAddr
         }
 
         friend class CSubNet;
+        friend class CNetAddrHash;
```
nit if you retouch, inverse the order of these two `friend class` declarements (both for alphabetical order and code order of the two classes, below)

vasild replied:
Done

jonatack reviewed (commented):
Code review ACK 7dddf330fe9f17e20df7bc7431ad3f7b740a3a9c also: debug build is clean, unit tests are green, ran bitcoind and observed debug log and -netinfo

fanquake reviewed (commented):
Concept ACK. I think we can merge this after the next branch off, once we start requiring C++17, and skip the addition of the pre-C++14 code.

jnewbery commented:
concept ACK. Will review after split-off when the pre-C++14 code is removed.

vasild commented:
Rebased to resolve conflicts.

vasild commented:
Addressed a [nit](https://github.com/bitcoin/bitcoin/pull/18722#discussion_r487708556) and removed pre-C++14 code.

This should be merged only after `-std=c++17` becomes mandatory.

vasild commented:
> ... after `-std=c++17` becomes mandatory

it is now

vasild commented:
d517c9d37...6f0c9f539: rebase due to conflicts

vasild commented:
`6f0c9f5390...6a1a28f2ad`: rebased due to conflicts, plus add `m_net` to the `CNetAddr` hashing.

sipa commented on src/netaddress.h:
```diff
@@ -477,6 +480,22 @@ class CNetAddr
+    size_t operator()(const CNetAddr& a) const noexcept
+    {
+        CSipHasher hasher(m_salt_k0, m_salt_k1);
+        hasher.Write(reinterpret_cast<const uint8_t*>(&a.m_net), sizeof(a.m_net));
```
Micronit: this reinterpret cast is a bit ugly. You could avoid it by just doing `hasher.Write(a.m_net);`; it'll interpret `a.m_net` as a uint64_t and use the optimized integer hasher.

vasild replied:
Done. I hope the implicit conversion from `int` to `uint64_t` will not upset some compiler.

jonatack replied:
The ugliness could maybe be considered a virtue; it's explicit that this is an *"on your head be it!"* type of cast and highlights that the code could be improved. https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast

vasild replied:
Here we can do without a cast since there is `Write()` method that takes an integer argument (modulo `int` -> `uint64` conversion which is ok because we never pass negative values).

sipa commented:
utACK 6a1a28f2ad0d84bf30debbe94c5bd8db80a45702

vasild commented:
`6a1a28f2ad...c22e98b30d`: address review suggestion

jonatack reviewed (commented):
ACK c22e98b30d9cd5d84ee3c14488defc25b3f998dd

hebasto commented on src/addrman.h:
```diff
@@ -569,13 +572,13 @@ friend class CAddrManTest;
         // Prune new entries with refcount 0 (as a result of collisions).
         int nLostUnk = 0;
-        for (std::map<int, CAddrInfo>::const_iterator it = mapInfo.begin(); it != mapInfo.end(); ) {
+        for (MapInfo::const_iterator it = mapInfo.begin(); it != mapInfo.end(); ) {
```
nit:
```suggestion
        for (auto it = mapInfo.cbegin(); it != mapInfo.cend(); ) {
```

vasild replied:
Done. Also, skipping the mention of `MapInfo` here makes it used in just one place. Dropped it and used `std::unordered_map<...` directly since there is no repetition of the tedious `std::unordered_map<...` now. Thanks!

hebasto reviewed (approved):
ACK c22e98b30d9cd5d84ee3c14488defc25b3f998dd

jonatack commented on src/addrman.h:
```diff
@@ -435,13 +435,13 @@ class CAddrMan
-                std::map<int, CAddrInfo>::const_iterator itCopy = it++;
+                auto itCopy = it++;
                 Delete(itCopy->first);
                 nLostUnk++;
```
nit
```diff
-                auto itCopy = it++;
+                const auto itCopy = it++;
                 Delete(itCopy->first);
-                nLostUnk++;
+                ++nLostUnk;
```

vasild replied:
Done

jonatack reviewed (commented):
ACK 1f4ee9c90531986f7f045f1987df8b3a8fa60433

vasild commented:
`1f4ee9c905...a92485b2c2`: address review suggestion

jonatack commented:
ACK a92485b2c250fd18f55d22aa32722bf52ab32bfe

hebasto reviewed (approved):
re-ACK a92485b2c250fd18f55d22aa32722bf52ab32bfe, only suggested changes and rebased since my [previous](https://github.com/bitcoin/bitcoin/pull/18722#pullrequestreview-666663681) review.

achow101 commented:
ACK a92485b2c250fd18f55d22aa32722bf52ab32bfe
